### ISSUE TITLE
2 != 3  Why is that so difficult to understand?!?

### ISSUE THREAD

**Issue by cclauss** [caskroom/homebrew-cask#15604]
* Python2 and Python3 are *not* the same thing.
* Anaconda2 and Anaconda3 are *not* the same thing.
* Anaconda2 **v2.4.0** != Anaconda3 **v2.4.0** share the same version number but are **not** the same thing.

**When two files are the same thing then they have the same `shasum -a 256`.**  If two files to not have the same checksum then they are **not** the same thing.  This nonsense of _let's not accept that version numbers are important_ is making brew cask harder to understand and less useful to the average user.

To anyone in the Python community today it is **laughable to say that Python2 and Python3 are the same thing**.  They are widely known to be incompatible with one another.  Yet brew cask wants to only have one Python which is a bit too much nirvana for today's reality.  This brew cask community seems to want me to just trust that `brew cask install python` is going to magically install the version that I had intending to use on my machine.  P2 has the lionshare of the current market but P3 is the (incompatible) present and future but brew case knows exactly which Python is best for me.  The Python community knows that P2 != P3  and it is naive to hide that distinction between the two from the brew cask user.

The same problem extends to the Anaconda distribution of Python (now the **worlds most broadly used platform for Data Science**).  This community want me to just trust that `brew cask install anaconda` is going to magically install the version that I had been intending to use.  When I use the normal anaconds2 v2.4.0 distribution script to install anaconda2 then a ~/anaconda**2** directory is created for me.  If I use the normal anaconda3 v2.4.0 distribution script to install anaconda**3** then ~/anaconda3 directory for me.  None of the anaconda apps work after `brew cask install anaconda` because all version numbers are a bad and should be ignored.  The normal distribution makes a distinction between anaconda2 and anaconda3 but the brew cask community seems unwilling to honor that distinction.  Therefore nothing works.  https://github.com/caskroom/homebrew-cask/pull/15596

**Comment by jawshooah** [comment]
Whoa there, cowboy. Where are you getting the idea that we don't care about versions? We have an [entire tap](https://github.com/caskroom/homebrew-versions) dedicated to preserving alternate versions of casks that are worth keeping around. [anaconda2 is there](https://github.com/caskroom/homebrew-versions/blob/master/Casks/anaconda2.rb). What's the problem?

**Comment by adityadalal924** [comment]
Thank you for making an issue.

`brew cask install python` and `brew cask install python3` were (arguably) mistakes, and will be adressed in an upcoming cleanup. From that point onwards, you can use `brew install python` and `brew install python3`, which seems to fit in line with your expectations (i.e. explicitly define python version).

We are also cleaning up `caskroom/versions` because over time, [things like this happen](https://github.com/caskroom/homebrew-versions/pull/1478/files), and we are not a version management tool (You can use conda or pyenv).

So, it comes down to, what are you asking for? 

Do you want `brew cask install anaconda2` and `brew cask install anaconda3`, or are you looking for something else?

**Comment by vitorgalvao** [comment]
@jawshooah If you see #15596, you’ll get the context.

Clearly, we’re “not all good” (a reply that has since been deleted, it seems), but lets not dwell into that. I am, however, frankly getting a bit tired of having the people who contribute less be the people who complain the most when we don’t do their bidding, and act as if their personal opinion on how to manage the project should be followed to the letter.

Everyone wants things done their way, and it is impossible to satisfy everyone and keep any degree of sanity in the project.

Lets be clear, here: we do not work for you, or any individual in particular. Similarly, we’re not doing things just to annoy you (or anyone). This project was an organisational mess, and we know from experience how harder that makes our work and therefore hurts the project and its users. We also know that the only way we could get things back on the right track, was by implementing clear rules of how things should be organised. Thanks to that, this main repo has flourished once again (though there are still some hard edges).

We can now move to the other big organisational issue, [caskroom/versions](https://github.com/caskroom/homebrew-versions). That repo was always a dump. We’re currently discussing and trying to implement rules on how to make it presentable. To do that, we need to delete a lot of casks and move a bunch of others around. Does that mean that are going to be some bumps? Definitely. Are we doing this to try and make it all better as a whole? You bet!

You’re making a bunch of assumptions with your post, most of them wrong, and most of them derived of how you *think* we should operate, rather than how we do operate. That is, understandably, an easy mistake to make, but it in no way justifies your entitled tone.

Since this has been explained repeatedly to you and you insist on hammering the same point, I’ll try to, once more, make it clear to you. Please try to not be blinded by what you want, and try to understand how we manage these.

We care about versions, as evidenced by the fact we have a whole repo dedicated to them, as @jawshooah. No one claimed different python versions are the same thing. In addition, no one claimed both versions of Anaconda are the same thing. *They are not the same thing, but they are related*. I’ll say that again, **they are related**. Google Chrome and Google Chrome Canary are not the same thing, but they are related to each other; [spyder](https://pythonhosted.org/spyder/) for python 2 and spyder for python 3 are not the same thing, but they are related to each other; Google Chrome and Spyder are not related to each other; Firefox and Alfred are not related to each other. Amaconda2 and Anaconda3 are even presented on the same page right next to each other, enforcing how they are not the same, but are indeed related.

Now, the way we deal with *related* software is to put it in the versions repo. This means only one stays in the main repo. **One**, not two or more, one. At the moment we are somewhat inconsistent about this in some cases because we were not yet able to correct every case, but we’re getting there.

The homebrew-cask doesn’t want you to believe in magic (fine if you do, but we’ll not enforce that belief). The homebrew-cask community is also strongly positioned as not a discoverability service (oh, how many times we’ve discussed that in the past, search for those two words together and you’ll find a lot of threads) which means *you*, the user, are responsible for knowing what you are installing. Different versions of `gitx` ([`gitx`](https://github.com/caskroom/homebrew-cask/blob/ccaf00a189cdbe50109ba6bf294c2b72aecba784/Casks/gitx.rb), [`laullon-gitx`](https://github.com/caskroom/homebrew-cask/blob/ccaf00a189cdbe50109ba6bf294c2b72aecba784/Casks/laullon-gitx.rb), [`rowanj-gitx`](https://github.com/caskroom/homebrew-cask/blob/ccaf00a189cdbe50109ba6bf294c2b72aecba784/Casks/rowanj-gitx.rb))? Not much we can do about those existing and having the same name (those are related, but in a different sense, as they’re forks and have different maintainers), but we try to make them all available as we can. Did you want `rowanj-gitx` but blindly installed `gitx`? Check first! We can’t guess what you mean, and can only try to make things clear.

“But somethings aren’t very clear”, you might say. Yes, and in those cases usually that is due to our very strict [token reference](https://github.com/caskroom/homebrew-cask/blob/master/doc/cask_token_reference.md). “Why not loosen than up a bit?”, you might wonder. Because without it, the repos become unmanageable with duplicates. We know, we’ve been there and it was dreadful.

In the end, what we know is we need clear rules that are applied consistently.

I’m also getting a bit tired of pointing this out, but homebrew gives you access to [tap](https://github.com/Homebrew/homebrew/blob/master/share/doc/homebrew/brew-tap.md)s, which are a wonderful system. We support them as well. **You are not tied to us or our rules**. Strongly disagree? Make your own tap and use it. It’s incredibly easy to set up and use.

Now, I do believe I was respectful, cordial, clear, and through (also verbose, which is not uncommon) in this reply. If you, my fellow maintainers, or any other member of the community disagrees, please do let me know. I do try to always be fair and write only with good intentions, but sometimes that can be hard to transpire in writing.

**Comment by cclauss** [comment]
Thanks much for the passion and humor of your explaination.  (I do not believe in magic which is probably why I get cranky.). Things are much clearer to me now that I understand the reasons behind your approach.  It was a tough night last night and I am sorry that I vented at you guys.  I wrote the "We are all good." While the install was going on but deleted it when I saw that no anaconda apps launch.  I will try agin this morning to get anaconda working and will open another PR if I find any issues in the formula.  Thanks and keep up the good work.
